src/deploy/functions/iambinder.ts

```diff
@@ -0,0 +1,298 @@
+    for (const { role, members, condition } of bindings) {
+      if (!isRole(role)) {
+        throw new Error(`Invalid role. Role ${role} must be prefixed with "roles/"`);
+      }
+      bindingInternal.add(role, members, condition);
+    }
+    return bindingInternal;
+  }
+}
+
+/**
+ * IamBinder generates necessary IAM binding given different types of function endpoints.
+ */
+export class IamBinder {
+  additions: Record<Resource, IamBindings>; // resource -> bindings
+  storageServiceAccount: string | undefined;
+
+  constructor(private readonly projectId: string, private readonly projectNumber: string) {
+    this.additions = {};
+  }
+
+  async apply(): Promise<iam.Policy[]> {
+    const updates = Object.entries(this.additions).map(async ([resource, bindings]) => {
+      return await IamBinder.updatePolicy(resource as Resource, bindings);
+    });
+    return Promise.all(updates);
+  }
+
+  async addEndpoints(endpoints: backend.Endpoint[]): Promise<void> {
+    for (const endpoint of endpoints) {
+      await this.addEndpoint(endpoint);
+    }
+  }
+
+  private async addEndpoint(endpoint: backend.Endpoint): Promise<void> {
+    let serviceAccount = this.defaultComputeServiceAccount(endpoint);
+    if (endpoint.serviceAccountEmail && endpoint.serviceAccountEmail !== "default") {
+      serviceAccount = `serviceAccount:${endpoint.serviceAccountEmail}`;
+    }
+
+    for (const secret of endpoint.secretEnvironmentVariables || []) {
+      this.addBinding(
+        `//secretmanager.googleapis.com/projects/${this.projectNumber}/secrets/${secret.secret}`,
+        "roles/secretmanager.secretAccessor",
+        [serviceAccount]
+      );
+    }
+
+    if (backend.isEventTriggered(endpoint)) {
+      if (endpoint.platform === "gcfv2") {
+        // Eventarc requires pubsub service account to have serviceAccountTokenCreator role.
+        // See https://cloud.google.com/eventarc/docs/roles-permissions#triggertypes-roles.
+        this.addBinding(
+          `//cloudresourcemanager.googleapis.com/projects/${this.projectNumber}`,
+          "roles/iam.serviceAccountTokenCreator",
+          [`serviceAccount:service-${this.projectNumber}@gcp-sa-pubsub.iam.gserviceaccount.com`]
+        );
+        // Service account associated w/ eventarc trigger must have permission to invoke a Cloud Run instance.
+        // We don't allow users to specify this service account hence it will always be the default compute service
+        // account. Usually, the default service account has EDITOR role which includes permission to invoke a Cloud
+        // Run instance, but given strict org policy it may not have this permission.
+        // See https://cloud.google.com/compute/docs/access/service-accounts#default_service_account.
+        this.addBinding(
+          `//cloudresourcemanager.googleapis.com/projects/${this.projectNumber}`,
+          "roles/run.invoker",
+          [this.defaultComputeServiceAccount(endpoint)]
+        );
```

Comment:
Sigh I feel pretty sad about that - in theory, this limitation applies to any service-specific service account that's created when we enable services.

Reply:
I guess this is a more general question about our UX. @inlined - Should we try to get the project to the correct state by enabling role permissions manually rather than erroring and having the user re-try?

Reply:
If possible we should succeed on first deploy.